### PR TITLE
[feature]: support ssd_3d_head in 3DSSD

### PR DESCRIPTION
**Pull request by encore-zhou**
Notice
1. Add ssd_3d_head
2. Add unittest for ssd_3d_head
3. Add Anchorfreebboxcoder
4. Add configs for 3dssdnet
5. Refactor `votehead` and `votemodule`, and modify the related unittest

### CODE REVIEW

File: mmdet3d/models/dense_heads/ssd_3d_head.py [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+
+        Returns:
+            dict: Predictions of vote head.
+        """
+        seed_points = feat_dict['sa_xyz']
+        seed_features = feat_dict['sa_features']
+        seed_indices = feat_dict['sa_indices']
+
+        # 1. generate candidate points from seed_points
+        candidate_offset = self.candidate_points_layers(
```
Comment (xavierwu95) [inline mmdet3d/models/dense_heads/ssd_3d_head.py:193]: candidate_points_layers is similar to VoteModule. consider to replace it.

File: tests/test_heads.py [hunk 2]
```diff
@@ -658,3 +658,60 @@ def test_h3d_head():
...
+    # test get_boxes
+    points = torch.stack(points, dim=0)
+    results = self.get_bboxes(points, ret_dict, img_metas)
+    assert results[0][0].tensor.shape[0] >= 0
+    assert results[0][0].tensor.shape[1] == 7
+    assert results[0][1].shape[0] >= 0
+    assert results[0][2].shape[0] >= 0
+
+
+test_ssd_3d_head()
```
Comment (xavierwu95) [inline tests/test_heads.py:717]: delete test code

File: mmdet3d/core/bbox/coders/anchor_free_bbox_coder.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
+import numpy as np
+import torch
+
+from mmdet.core.bbox.builder import BBOX_CODERS
+from .partial_bin_based_bbox_coder import PartialBinBasedBBoxCoder
+
+
+@BBOX_CODERS.register_module()
+class AnchorFreeBBoxCoder(PartialBinBasedBBoxCoder):
```
Comment (xavierwu95) [inline mmdet3d/core/bbox/coders/anchor_free_bbox_coder.py:9]: consider to merge AnchorFreeBBoxCoder into PartialBinBasedBBoxCoder instead of inheriting it. use `if self.with_...` to control operation in encode, decode and split_pred.

File: configs/_base_/models/3dssd.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        vote_loss=dict(type='SmoothL1Loss', reduction='sum', loss_weight=1.5)))
+# model training and testing settings
+train_cfg = dict(pos_distance_thr=10.0, expand_dims_length=0.1)
+test_cfg = dict(
+    nms_thr=0.1,
+    score_thr=0.1,
+    per_class_proposal=True,
+    max_translate_range=(3.0, 3.0, 2.0))
+
+# optimizer
```
Comment (ZwwWayne) [inline configs/_base_/models/3dssd.py:59]: Line 59 - 67 should be moved to `3dssd/3dssd_kitti-3d-car.py`

File: mmdet3d/core/evaluation/kitti_utils/eval.py [hunk 5]
```diff
@@ -473,8 +473,9 @@ def eval_class(gt_annos,
     assert len(gt_annos) == len(dt_annos)
     num_examples = len(gt_annos)
     if num_examples < num_parts:
-        num_parts = num_examples
+        num_parts = 1
```
Comment (ZwwWayne) [inline mmdet3d/core/evaluation/kitti_utils/eval.py:476]: Is this change necessary?

Comment (encore-zhou) [reply mmdet3d/core/evaluation/kitti_utils/eval.py:476]: The original implementation may trigger error while `num_examples` smaller than `num_parts`

File: mmdet3d/core/post_processing/box3d_nms.py [hunk 6]
```diff
@@ -134,3 +134,47 @@ def aligned_3d_nms(boxes, scores, classes, thresh):
 
     indices = boxes.new_tensor(pick, dtype=torch.long)
     return indices
+
+
+def aligned_bev_nms(boxes, scores, classes, thresh):
```
Comment (ZwwWayne) [inline mmdet3d/core/post_processing/box3d_nms.py:139]: Can we directly use the 2D NMS implemented in mmcv by passing only the 4 necessary dimensions?

File: configs/_base_/models/3dssd.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+train_cfg = dict(
+    sample_mod='spec', pos_distance_thr=10.0, expand_dims_length=0.05)
+test_cfg = dict(
+    sample_mod='spec',
+    nms_thr=0.1,
+    score_thr=0.0,
+    per_class_proposal=True,
+    max_output_num=100)
+
+# optimizer
```
Comment (ZwwWayne) [inline configs/_base_/models/3dssd.py:80]: Move line 80-88 to 3dssd_kitti-3d-car.py

File: mmdet3d/models/backbones/pointnet2_sa_msg.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
+import torch
```
Comment (ZwwWayne) [inline mmdet3d/models/backbones/pointnet2_sa_msg.py:1]: Please merge the newest master and clean the duplicated codes.

File: mmdet3d/models/backbones/pointnet2_sa_msg.py [hunk 9]
```diff
@@ -103,7 +106,7 @@ def __init__(self,
                     sa_out_channel,
                     aggregation_channels[sa_index],
                     conv_cfg=dict(type='Conv1d'),
-                    norm_cfg=dict(type='BN1d'),
+                    norm_cfg=dict(type='BN1d', eps=1e-3),
```
Comment (ZwwWayne) [inline mmdet3d/models/backbones/pointnet2_sa_msg.py:109]: Put the eps argument in the config rather than in the default arguments. Also, if it is not necessary, we do not need use them.

File: mmdet3d/models/dense_heads/mlp_bbox_head.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    """
+
+    def __init__(self,
+                 in_channels=0,
+                 shared_mlp_channels=(),
+                 cls_mlp_channels=(),
+                 num_cls_out_channels=0,
+                 reg_mlp_channels=(),
+                 num_reg_out_channels=0,
+                 conv_cfg=dict(type='Conv1d'),
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/mlp_bbox_head.py:26]: If it is Conv1d here, maybe we could not call it as mlpbboxhead. We may use `BaseAnchorFree3DHead`. And use it as a base for VoteHead & SSD3DHead.

File: mmdet3d/models/dense_heads/mlp_bbox_head.py [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+            out_channels = self.in_channels
+
+        # add cls specific branch
+        prev_channel = out_channels
+        if len(self.cls_mlp_channels) > 0:
+            self.cls_mlps = self._add_mlp_branch(prev_channel,
+                                                 self.cls_mlp_channels)
+            prev_channel = self.cls_mlp_channels[-1]
+
+        self.cls_pred = nn.Conv1d(prev_channel, num_cls_out_channels, 1)
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/mlp_bbox_head.py:62]: Use `build_conv_layer` API with conv_cfg here thus we could also use `Conv2d` in this head.

File: mmdet3d/models/dense_heads/ssd_3d_head.py [hunk 12]
```diff
@@ -0,0 +1,545 @@
...
+        else:
+            raise NotImplementedError('Unsupported bbox type!')
+
+        corner3d = bbox.corners
+        minmax_box3d = corner3d.new(torch.Size((corner3d.shape[0], 6)))
+        minmax_box3d[:, :3] = torch.min(corner3d, dim=1)[0]
+        minmax_box3d[:, 3:] = torch.max(corner3d, dim=1)[0]
+
+        bbox_classes = torch.argmax(sem_scores, -1)
+        nms_selected = aligned_bev_nms(minmax_box3d[nonempty_box_mask],
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/ssd_3d_head.py:486]: I suggest directly use the nms API implemented in MMCV.

File: mmdet3d/models/model_utils/vote_module.py [hunk 13]
```diff
@@ -88,21 +115,37 @@ def forward(self, seed_points, seed_feats):
...
-                                     feat_channels).transpose(2,
-                                                              1).contiguous()
-
-        if self.norm_feats:
-            features_norm = torch.norm(vote_feats, p=2, dim=1)
-            vote_feats = vote_feats.div(features_norm.unsqueeze(1))
-        return vote_points, vote_feats
+        offset = offset.reshape(batch_size, num_vote, 3).transpose(2, 1)
+
+        if self.with_feat_residual:
```
Comment (ZwwWayne) [inline mmdet3d/models/model_utils/vote_module.py:134]: self.with_feat_residual -> self.with_res_feat

File: configs/3dssd/3dssd_kitti-3d-car.py [hunk 14]
```diff
@@ -0,0 +1,152 @@
+_base_ = ['../_base_/models/3dssd.py', '../_base_/default_runtime.py']
```
Comment (ZwwWayne) [inline configs/3dssd/3dssd_kitti-3d-car.py:1]: Inherits the kitti-3d-car config in base.

File: mmdet3d/models/model_utils/vote_module.py [hunk 15]
```diff
@@ -23,24 +25,38 @@ class VoteModule(nn.Module):
...
                  vote_per_seed=1,
                  gt_per_seed=3,
+                 num_points=-1,
                  conv_channels=(16, 16),
                  conv_cfg=dict(type='Conv1d'),
                  norm_cfg=dict(type='BN1d'),
+                 act_cfg=dict(type='ReLU'),
                  norm_feats=True,
+                 with_res_feat=True,
+                 clip_xyz_offset=None,
```
Comment (ZwwWayne) [inline mmdet3d/models/model_utils/vote_module.py:45]: clip_xyz_offset -> vote_xyz_range

File: mmdet3d/models/dense_heads/base_conv_bbox_head.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+            out_channels=num_reg_out_channels,
+            kernel_size=1)
+
+    def _add_mlp_branch(self, in_channels, mlp_channels):
+        """Add shared or separable branch."""
+        mlp_spec = [in_channels] + list(mlp_channels)
+        # add branch specific mlp layers
+        mlp = nn.Sequential()
+        for i in range(len(mlp_spec) - 1):
+            mlp.add_module(
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/base_conv_bbox_head.py:87]: Change name mlp to conv_layers

File: mmdet3d/models/dense_heads/base_conv_bbox_head.py [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+            out_channels = self.in_channels
+
+        # add cls specific branch
+        prev_channel = out_channels
+        if len(self.cls_mlp_channels) > 0:
+            self.cls_mlps = self._add_mlp_branch(prev_channel,
+                                                 self.cls_mlp_channels)
+            prev_channel = self.cls_mlp_channels[-1]
+
+        self.cls_pred = build_conv_layer(
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/base_conv_bbox_head.py:63]: self.cls_pred -> self.conv_cls
self.reg_pred -> self.conv_reg

File: mmdet3d/models/dense_heads/base_conv_bbox_head.py [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+                                                 self.reg_mlp_channels)
+            prev_channel = self.reg_mlp_channels[-1]
+
+        self.reg_pred = build_conv_layer(
+            conv_cfg,
+            in_channels=prev_channel,
+            out_channels=num_reg_out_channels,
+            kernel_size=1)
+
+    def _add_mlp_branch(self, in_channels, mlp_channels):
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/base_conv_bbox_head.py:81]: _add_conv_branch

File: mmdet3d/models/dense_heads/base_conv_bbox_head.py [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+
+        self.cls_pred = build_conv_layer(
+            conv_cfg,
+            in_channels=prev_channel,
+            out_channels=num_cls_out_channels,
+            kernel_size=1)
+        # add reg specific branch
+        prev_channel = out_channels
+        if len(self.reg_mlp_channels) > 0:
+            self.reg_mlps = self._add_mlp_branch(prev_channel,
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/base_conv_bbox_head.py:71]: self.reg_mlps -> self.reg_convs

File: mmdet3d/models/dense_heads/base_conv_bbox_head.py [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+        self.reg_mlp_channels = reg_mlp_channels
+        self.num_reg_out_channels = num_reg_out_channels
+        self.conv_cfg = conv_cfg
+        self.norm_cfg = norm_cfg
+        self.act_cfg = act_cfg
+        self.bias = bias
+
+        # add shared convs and fcs
+        if len(self.shared_mlp_channels) > 0:
+            self.shared_mlps = self._add_mlp_branch(self.in_channels,
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/base_conv_bbox_head.py:50]: self.shared_convs

File: mmdet3d/models/dense_heads/base_conv_bbox_head.py [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+                                                    self.shared_mlp_channels)
+            out_channels = self.shared_mlp_channels[-1]
+        else:
+            out_channels = self.in_channels
+
+        # add cls specific branch
+        prev_channel = out_channels
+        if len(self.cls_mlp_channels) > 0:
+            self.cls_mlps = self._add_mlp_branch(prev_channel,
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/base_conv_bbox_head.py:59]: self.cls_mlps -> self.cls_convs

File: configs/_base_/models/3dssd.py [hunk 22]
```diff
@@ -42,9 +42,9 @@
             bias=True),
         pred_layer_cfg=dict(
             in_channels=1536,
-            shared_mlp_channels=(512, 128),
-            cls_mlp_channels=(128, ),
-            reg_mlp_channels=(128, ),
+            shared_conv_layers=(512, 128),
```
Comment (ZwwWayne) [inline configs/_base_/models/3dssd.py:45]: shared_conv_layers -> shared_conv_channels

File: mmdet3d/models/dense_heads/base_conv_bbox_head.py [hunk 23]
```diff
@@ -67,36 +67,36 @@ def __init__(self,
...
+    def _add_conv_branch(self, in_channels, mlp_channels):
         """Add shared or separable branch."""
-        mlp_spec = [in_channels] + list(mlp_channels)
-        # add branch specific mlp layers
-        mlp = nn.Sequential()
-        for i in range(len(mlp_spec) - 1):
-            mlp.add_module(
+        conv_spec = [in_channels] + list(mlp_channels)
+        # add branch specific conv layers
+        conv = nn.Sequential()
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/base_conv_bbox_head.py:85]: conv -> conv_layers or convs

**Review by ZwwWayne — APPROVED** [review]
